src/Arcus.Security.Core/CompositeSecretProvider.cs

```diff
@@ -127,70 +139,74 @@ public ISecretProvider GetProvider(string name)
         public ICachedSecretProvider GetCachedProvider(string name)
         {
             Guard.NotNullOrWhitespace(name, nameof(name), "Requires a non-blank name to retrieve the registered named secret provider");
-
-            SecretStoreSource source = GetSecretSource(name);
-            if (source.CachedSecretProvider is null)
+            Guard.For<NotSupportedException>(
+                () => !HasCachedSecretProviders, 
+                $"Cannot use cached secret store operation because none of the secret providers in the secret store were registered as cached secret providers ({nameof(ICachedSecretProvider)})");
+            
+            ISecretProvider source = GetSingleOrSubsetSecretProvider(name);
+            if (source is ICachedSecretProvider cachedSource)
             {
-                throw new NotSupportedException(
-                    $"Found a registered {nameof(ISecretProvider)} with the name '{name}' in the secret store, but was not configured for caching. "
-                    + $"Please use the {nameof(GetProvider)} instead or configure the registered provider with caching");
+                return cachedSource;
             }
 
-            return source.CachedSecretProvider;
+            throw new NotSupportedException(
+                $"Found a registered {nameof(ISecretProvider)} with the name '{name}' in the secret store, but was not configured for caching. "
+                + $"Please use the {nameof(GetProvider)} instead or configure the registered provider with caching");
         }
 
-        private SecretStoreSource GetSecretSource(string name)
+        private ISecretProvider GetSingleOrSubsetSecretProvider(string name)
         {
-            IEnumerable<SecretStoreSource> matchingProviders =
-                _secretProviders.Where(provider => provider.Options?.Name == name);
-
-            int count = matchingProviders.Count();
-            if (count is 0)
-            {
-                throw new KeyNotFoundException(
-                    $"Could not retrieve the named {nameof(ISecretProvider)} because no provider was registered with the name '{name}'");
-            }
-
-            if (count > 1)
+            if (_groupedSecretStores.TryGetValue(name, out Lazy<ISecretProvider> subsetSecretStore))
             {
-                throw new KeyNotFoundException(
-                    $"Could not retrieve the named {nameof(ISecretProvider)} because more than one provider was registered with the name '{name}'");
+                return subsetSecretStore.Value;
             }
-
-            SecretStoreSource firstProvider = matchingProviders.First();
-            return firstProvider;
+            
+            throw new KeyNotFoundException(
+                $"Could not retrieve the named {nameof(ISecretProvider)} because no secret provider was registered with the name '{name}'");
```

Comment:
```suggestion

            throw new KeyNotFoundException(
                $"Could not retrieve the named {nameof(ISecretProvider)} because no secret provider was registered with the name '{name}'");
```